rootfs/etc/nginx/lua/balancer/ewma.lua

```diff
@@ -0,0 +1,128 @@
+-- Original Authors: Shiv Nagarajan & Scott Francis
+-- Accessed: March 12, 2018
+-- Inspiration drawn from:
+-- https://github.com/twitter/finagle/blob/1bc837c4feafc0096e43c0e98516a8e1c50c4421/finagle-core/src/main/scala/com/twitter/finagle/loadbalancer/PeakEwma.scala
+
+
+local resty_lock = require("resty.lock")
+local util = require("util.util")
+local split = require("util.split")
+
+local DECAY_TIME = 10 -- this value is in seconds
+local LOCK_KEY = ":ewma_key"
+local PICK_SET_SIZE = 2
+
+local ewma_lock, err = resty_lock:new("locks", {timeout = 0, exptime = 0.1})
+
+local _M = {}
+
+local function lock(upstream)
+  local _, err = ewma_lock:lock(upstream .. LOCK_KEY)
+  if err then
+    if err ~= "timeout" then
+      ngx.log(ngx.ERR, string.format("EWMA Balancer failed to lock: %s", tostring(err)))
+    end
+  end
+
+  return err
+end
+
+local function unlock()
+  local ok, err = ewma_lock:unlock()
+  if not ok then
+    ngx.log(ngx.ERR, string.format("EWMA Balancer failed to unlock: %s", tostring(err)))
+  end
+
+  return err
+end
+
+local function decay_ewma(ewma, last_touched_at, rtt, now)
+  local td = now - last_touched_at
+  td = (td > 0) and td or 0
+  local weight = math.exp(-td/DECAY_TIME)
+
+  ewma = ewma * weight + rtt * (1.0 - weight)
+  return ewma
+end
+
+local function get_or_update_ewma(upstream, rtt, update)
+  local lock_err = nil
+  if update then
+    lock_err = lock(upstream)
+  end
+  local ewma = ngx.shared.balancer_ewma:get(upstream) or 0
+  if lock_err ~= nil then
+    return ewma, lock_err
+  end
+
+  local now = ngx.now()
+  local last_touched_at = ngx.shared.balancer_ewma_last_touched_at:get(upstream) or 0
+  ewma = decay_ewma(ewma, last_touched_at, rtt, now)
+
+  if not update then
+    return ewma, nil
+  end
+
+  ngx.shared.balancer_ewma_last_touched_at:set(upstream, now)
+  ngx.shared.balancer_ewma:set(upstream, ewma)
+
+  unlock()
+  return ewma, nil
+end
+
+
+local function score(upstream)
+  -- Original implementation used names
+  -- Endpoints don't have names, so passing in IP:Port as key instead
+  local upstream_name = upstream.address .. upstream.port
```

Comment:
If you look at https://github.com/Shopify/ingress/pull/19/files#diff-e50d7e0b064eb1aa4de48772ee22f13cR119 it uses `upstream_addr` var from Nginx which has the upstreams as `<ip>:<port>`, therefore I think this should be concatenated using `:` as well.